arroyo-console/src/routes/sources/SourceDetail.tsx

```diff
@@ -0,0 +1,141 @@
+                    newOutputs.shift();
+                }
+                return newOutputs;
+            });
+        }
+        setSubscribed(false);
+    }
+
+    const blockUntilJobReady = async () => {
+        let job = await (await client()).getJobDetails(new JobDetailsReq({ jobId: jobId! }));
+        let jobState = job.jobStatus?.state;
+        setJobState(jobState);
+        if (job.jobStatus?.state != "Running") {
+            await new Promise(r => setTimeout(r, 1000));
+            await blockUntilJobReady();
+        }
+    }
+
+    useEffect(() => {
+        if (jobId != null) {
+            setSubscribed(true);
+            blockUntilJobReady().then(subscribe);
+        }
+    }, [jobId]);
+
+    return (
+        <Box>
+            <PipelineOutputs outputs={outputs} />
+        </Box>
+    )
+}
+
+function DetailFieldTable({ data }: { data: Array<{ field: string, value: string }> }) {
+    return (
+        <TableContainer w="400px">
+            <Table variant="unstyled" bg="">
+                <Tbody>
+                    {data.map((row, i) => {
+                        if (row.value != "") {
+                            return (
+                                <Tr key={i}>
+                                    <Td textAlign="right" w="15%" fontWeight="bold" fontSize="sm">{row.field}</Td>
+                                    <Td textAlign="left" fontSize="xl">{row.value}</Td>
+                                </Tr>
+                            )
+                        }
+                    })}
+                </Tbody>
+            </Table>
+        </TableContainer>
+    )
+}
+
+export function SourceDetail({ client }: { client: ApiClient }) {
+    let [source, setSource] = useState<SourceDef | null>(null);
+    let [previewJobId, setPreviewJobId] = useState<string | null>(null);
+    let [subscribed, setSubscribed] = useState(false);
+    let [jobState, setJobState] = useState<string | undefined>(undefined);
+
+
+    let { id } = useParams();
+
+    useEffect(() => {
+        // initialize the source given the ID
+        const fetchData = async () => {
+            const source = await (await client()).getSource({ id: BigInt(id!) });
+
+            setSource(source.source!);
+        }
+        fetchData()
+    }, []);
+
+    // log a message when subscribed changes
+    useEffect(() => {
+        console.log(`subscribed: ${subscribed}`);
+    }, [subscribed]);
+
+    const tail = async () => {
+        let jobResp = await (await client()).previewPipeline(new CreatePipelineReq({
+            name: `preview-source-${source!.name}`,
+            config: {
+                case: "sql",
+                value: new CreateSqlJob({
+                    query: `SELECT * FROM ${source!.name}; `,
+                    sink: { case: "builtin", value: BuiltinSink.Web }
+                }),
+            },
+        }));
+        setPreviewJobId(jobResp.jobId);
+    }
+
+    return (
+        <Box>
```

Comment:
The spacing of this is a bit inconsistent with the rest of the UI, and having the badge and the button next to each other makes it somewhat unclear which one is the button.

 I'd recommend something like this for the header:

```tsx
      <VStack>
        <Flex width="100%" p={5}>
          <Heading size="sm">Source {source?.name} </Heading>
          <Center>{jobState != null ? <Badge ml={3}>{jobState}</Badge> : null}</Center>
          <Spacer />

          <Button size="sm" onClick={tail} isLoading={subscribed} loadingText="In Progress">
            Tail
          </Button>
        </Flex>
```